lib/core/proxy.c

```diff
@@ -319,8 +354,19 @@ static void do_close(struct rp_generator_t *self)
     }
     h2o_timer_unlink(&self->send_headers_timeout);
     if (self->pipe_reader.fds[0] != -1) {
-        close(self->pipe_reader.fds[0]);
-        close(self->pipe_reader.fds[1]);
+        h2o_conn_t *conn = self->src_req->conn;
+        if ((pipe_list.len < conn->ctx->globalconf->proxy.max_pipes) && empty_pipe(self->pipe_reader.fds[0])) {
```

Comment:
FWIW, we can determine if the pipe is empty by checking if `self->body_bytes_read == self->body_bytes_sent`. Iff the two variables are equal, the pipe is empty.

Hence I think the question would be:

1. Do we want to do something when the pipe is known to be empty? Maybe we do not. Or if we do want to, maybe we'd better do `read` and call `h2o_fatal` if we succeed in reading, because that means that we have inconsistency among the states.
2. When we know that the response was closed abruptly and that there is data in the pipe, do we want to drain it instead of just closing it? We can just close assuming that it would be rare to close abruptly, or drain to avoid the cost of calling `pipe` again soon.

Reply:
I'm going to mark this as resolved for now because we discussed the `h2o_fatal` being triggered during local testing with abrupt closure and I think you suggested we go with draining the pipe for now until further investigation.

since we are going to go with the original behavior of draining the fd for the time being, i'll mark this as resolved but we can change it in the future if needed/desired.